### ISSUE TITLE
Differences in XML parsing between Marlin and the converter script

### ISSUE THREAD

**Issue by tmadlener** [key4hep/k4MarlinWrapper#35]
It looks like there are some differences between the xml that is considered valid by `Marlin` and the xml that is considered valid by the `convertMarlinSteeringToGaudi.py` script. Here are two examples that I just encountered.

### Treating `&` as escape character or not
Considering the following [line from a steering file](https://github.com/iLCSoft/ILDConfig/blob/6bf49e0b7d36074ed0987d03b8721ba1c16e102d/StandardConfig/production/MarlinStdRecoMiniDST.xml#L330)
```xml
    <parameter name="FlavorTag.CategoryDefinition2" type="string">nvtx==1&&nvtxall==1</parameter>
```
`Marlin` seems to parse this just fine and work without any problems, but trying to convert the steering file results in a parsing error:
```console
Exception when getting trees: ParseError('not well-formed (invalid token): line 330, column 72') 
```
It looks like `Marlin` is a bit more permissive here, since `&` is an escape character in XML.

### Automatically replacing environment variables
From the same steering file, consider [this line](https://github.com/iLCSoft/ILDConfig/blob/6bf49e0b7d36074ed0987d03b8721ba1c16e102d/StandardConfig/production/MarlinStdRecoMiniDST.xml#L45)
```xml
  <constant name="lcgeo_DIR" value=$lcgeo_DIR/>
```
Here `Marlin` again happily replaces `$lcgeo_DIR` with the value of that environment variable but the converter script runs into a parsing error:
```console
Exception when getting trees: ParseError('not well-formed (invalid token): line 45, column 35')
```
I am not entirely sure if this replacement happens automatically within the xml parser used by `Marlin` or if this is done deliberately.

-----
This seems to be a continuation of #18 but I am not sure if a fix exists for these new ones here. They can be fixed by changing the original steering file and/or by parsing it with `Marlin` first to do some of the replacements and producing a parsed steering file from the original one and then passing the parsed one to the converter.

As "quick fixes" the following can be done:
- replace all `&&` with `&amp;&amp` in the original steering file
- Do the replacement of the environment variables yourself or use the `global.OutputSteeringFile` parameter and `Marlin -n --global.OutputSteeringFile=parsed.xml <other-constants-to-replace> steering-file.xml`. (Note that you might have to introduce the `OutputSteeringFile` parameter in the `global` section first to be able to override it)

**Comment by andresailer** [comment]
Probably in both cases (ampersands and environment variables) we can pre-process the xml file to something that is valid XML and post process it back to have the ampersands in the string and use os.environ to read the environment variables?

1. read xml 
1. replace amps and strings like '^\$[a-zA-Z0-9_]` (?) with something magic?
1. do xml to python conversion
1. replace magic back with os.environ, ampersands
2. write output python file

are environment variables only in the constants, or anywhere?

**Comment by tmadlener** [comment]
I have checked again for the environment variables and this seems to be something that is supported by the underlying xml parsing library of Marlin rather than Marlin itself. At least in a Key4hep environment the replacement of env variables in the `constant` section is no longer working, while in the latest iLCSoft release it does. So I am actually not sure if we need to support that in the end.

For the pre-processing of the amps we will probably have to find something. Where I stumbled over it, it is basically passed directly to a [`TTreeFormula`](https://root.cern.ch/doc/master/classTTreeFormula.html), which unfortunately is not very happy with simply replacing the `&&` with `and`.

Is there a way to do the pre-processing in memory? Or does the parsing only work from file?
